pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -338,15 +340,115 @@ func (pl *VolumeRestrictions) EventsToRegister() []framework.ClusterEventWithHin
 		// A new Node may make a pod schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdded},
 		// Pods may fail to schedule because the PVC it uses has not yet been created.
 		// This PVC is required to exist to check its access modes.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted,
+// It checks whether the deleted pod will conflict with volumes of other pods on the same node
+func (pl *VolumeRestrictions) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyReadWriteOncePodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
+	}
+
+	nodeInfo := framework.NewNodeInfo(oldPod)
+	if !satisfyVolumeConflicts(pod, nodeInfo) {
+		return framework.QueueAfterBackoff
+	}
+
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterNodeAdded is invoked whenever a node added, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterNodeAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, newNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyReadWriteOncePodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
+	}
+
+	logger.V(4).Info("node was added", "pod", klog.KObj(pod), "node", klog.KObj(newNode))
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterPersistentVolumeClaimChange is invoked whenever a PersistentVolumeClaim changed, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterPersistentVolumeClaimChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPersistentVolumeClaim, _, err := util.As[*v1.PersistentVolumeClaim](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPersistentVolumeClaimChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
```

Comment:
We should check if the added PVC is this Pod's one or not.

Reply:
I don't quite understand why we need to check if the pvc is for this pod, may you explain in detail?

According to the [Filter](https://github.com/kubernetes/kubernetes/blob/95c6295636adaeb06d47b02ff4c705b6aa106606/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L320C1-L331), we will only determine if the pod has a conflicting ReadWriteOncePodPVC

Let's assume the following scenario

* If we find that the new pvc is for this pod, it certainly won't conflict with other pods, so we think it has no effect on pod scheduling, so we should skip it

* If we find that the new pvc is not of this pod, it again will not conflict with other pods, so we assume it has no impact on pod scheduling, so we should skip it

Am I understanding this correctly?

Reply:
Apologize, I meant the next line L399 (`if oldPersistentVolumeClaim == nil`).

In PreFilter, we're returning unschedualble when PVC for this Pod doesn't exist.
https://github.com/sanposhiho/kubernetes/blob/408e45a66a82a566e4684d6a97f9810127b985f9/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L182

And, in that case, we want to queue the Pod when the PVC for this Pod is created.
But, currently, we queue the Pod when any PVC is added.

Reply:
I got it. Thanks for your explanation.